pylds/base.py

```diff
@@ -6,32 +6,122 @@
+
+def energy_conservation_condition(phase_space_axes, H0, potential_energy, momentum_sign):
+    N_dim = len(phase_space_axes)
+    phase_space_positions = np.transpose(phase_space_axes[:int(N_dim/2)])
+    phase_space_momenta = np.transpose(phase_space_axes[int(N_dim/2):])
+    V = potential_energy(phase_space_positions)
+    points_dof_remaining = momentum_sign * \
+        np.sqrt(2*(H0 - V) - (phase_space_momenta**2).sum(axis=1))
+    return points_dof_remaining
 
 def generate_points(grid_parameters):
     """
     Returns a 1D array of all points from a on a uniform grid with dimensions and size defined by list of input parameters.
     An additional dimension initiallised with zeros is added for the calculation of Lagrangian Descriptors.
-    
+
     Parameters
     ----------
-    grid_parameters : list of 3-tuples of floats
+    slice_parameters : list of 3-tuples of floats
         input parameters of limits and size of mesh per axis
-        
+
     Returns
     -------
     mesh : 1d numpy array
         flattened array of initial conditions
     """
-    x_min, x_max, Nx = grid_parameters[0]
-    y_min, y_max, Ny = grid_parameters[1]
-    points_x = np.linspace(x_min, x_max, Nx)
-    points_y = np.linspace(y_min, y_max, Ny)    
-    Y, X = np.meshgrid(points_y, points_x)  # Grid in phase-space
-    mesh = np.transpose([X.flatten(), Y.flatten(), np.zeros(Nx*Ny)]) # 2D grid + a zero column for LDs
-    return mesh.flatten()
+    if type(grid_parameters) == dict:
+        # Unpack extra grid parameters
+        slice_parameters = grid_parameters['slice_parameters']
+        dof_slice = grid_parameters['dof_slice']
+        dof_fixed, dof_fixed_values = grid_parameters['dof_fixed']
+        momentum_sign = grid_parameters['momentum_sign']
+        potential_energy = grid_parameters['potential_energy']
+        H0 = grid_parameters['energy_level']
+
+        N_dim = len(dof_slice)  # Total number of DoF
+        # Check N DoF is even.
+        if N_dim % 2 != 0:
+            error_mssg = ("ERROR: Number of DoF not even. ",
+                          "Check your extra grid parameters")
+            print(error_mssg)
+            sys.exit()
+        # Determine number of DoF for Energy conservation.
+        # There must be only one.
+        dof_remaining = 1 - (np.array(dof_fixed) + np.array(dof_slice))
+        if list(dof_remaining).count(1) > 1:
+            error_mssg = ("ERROR: More than one remaing DoF. ",
+                          "Cannot use Energy conservation to define high-dim grid.")
+            print(error_mssg)
+            sys.exit()
+
+        # Axes of visualisation slice
+        def pass_axis_parameters(parameters): return np.linspace(*parameters)
+        points_slice_axes = list(map(pass_axis_parameters, slice_parameters))
+        slice_mesh = np.meshgrid(*points_slice_axes)
+        dof_slice_axes = [axis.flatten() for axis in slice_mesh]
+
+        # Axes of fixed DoF
+        dof_fixed_axes = dof_fixed_values
```

Comment:
`dof_fixed_axes` can be defined from the beginning, as `dof_fixed_values` aren't used anywhere. (dof comment applicable)

Reply:
Yeah, good point. I will keep everything as `dof_fixed_values` then.